### ISSUE TITLE
[css-cascade-6] Is it possible to create a scoping boundary over sibling elements?

### ISSUE THREAD

**Issue by romainmenke** [w3c/csswg-drafts#7233]
https://www.w3.org/TR/2021/WD-css-cascade-6-20211221/#scoped-styles

> [Scoping limits](https://www.w3.org/TR/2021/WD-css-cascade-6-20211221/#scoping-limit) can use the [:scope](https://www.w3.org/TR/selectors-4/#scope-pseudo) pseudo-class to require a specific relationship to the [scoping root](https://www.w3.org/TR/selectors-4/#scoping-root):
> ```css
> /* .content is only a limit when it is a direct child of the :scope */
> @scope (.media-object) to (:scope > .content) { ... }
> ```
> [Scoping limits](https://www.w3.org/TR/2021/WD-css-cascade-6-20211221/#scoping-limit) can also reference elements outside their [scoping root](https://www.w3.org/TR/selectors-4/#scoping-root) by using [:scope](https://www.w3.org/TR/selectors-4/#scope-pseudo). For example:
> ```css
> /* .content is only a limit when the :scope is inside .sidebar */
> @scope (.media-object) to (.sidebar :scope .content) { ... }
> ```

The scoping proposal is mostly focussed on component based styling.

I think the weak scoping proximity feature that is part of the scoping proposal would also be useful for sibling selectors. I was not sure if this was also covered by the current specification.

`Example 3` seems to allow room for a scope with this definition :

```css
@scope (.media-object) to (:scope ~ .text-object) { ... }
```

---------

### Example :

Styling any paragraph elements that follow a certain heading element.

```css
h1 ~ p {
	color: red;
}

h2 ~ p {
	color: green;
}

h3 ~ p {
	color: blue;
}
```

```html
<h1>Main heading (h1)</h1>

<p>Red text</p>

<h2>Sub heading (h2)</h2>

<p>Green text</p>
<p>Green text</p>

<h3>Smaller heading (h3)</h3>

<p>Blue text</p>
<p>Blue text</p>

<h2>Sub heading (h2)</h2>

<p>Blue text, but intended green</p>
<p>Blue text, but intended green</p>
```

<img width="322" alt="Screenshot of example HTML and CSS described above showing how the general sibling selector in CSS also does not account for element proximity." src="https://user-images.githubusercontent.com/11521496/165245715-0a2815b7-a064-41c1-9b32-a0e7f20e6ce5.png">

Possibly we can write this CSS :

```css
@scope (h1) to (:scope ~ :is(h1, h2, h3, h4, h5, h6)) {
	p {
		color: red;
	}
}

@scope (h2) to (:scope ~ :is(h1, h2, h3, h4, h5, h6)) {
	p {
		color: green;
	}
}

@scope (h3) to (:scope ~ :is(h1, h2, h3, h4, h5, h6)) {
	p {
		color: blue;
	}
}
```

**Comment by romainmenke** [comment]
@mirisuzanne @fantasai any thoughts or suggestions for this case?

**Comment by mirisuzanne** [comment]
Hi @romainmenke, it's an interesting use-case - and it does require something similar to scope. I find that really interesting to consider, but I think it needs a bit of disentangling the parts. The current scope spec has two aspects, and their use-cases overlap:

- 'scope' itself is mostly about limiting the range of elements that can be selected. That's handled with a root and boundaries.
- 'proximity' gives cascade priority to the selector with a 'closer' relationship between elements in the DOM.

As far as I can tell, your example would only need one of those two features. If you weight the proximity of siblings, you no longer need sibling-boundaries (the 'scope' part), and if you 'scope' the selectors to a particular range of siblings (implied by you suggested syntax), you no longer need the proximity weighting (since the scopes don't overlap).

In order to define sibling-based 'scopes', we would need a more major shift to the syntax. Currently, the 'root' selector is able to generate a scope based on the DOM tree that includes the root and all its descendants. Then the boundaries are scoped to that tree, narrowing it further, and nested selectors only match in the resulting subtree fragment. For a sibling-based version, the root selector would have to take on an entirely different meaning, returning potentially multiple sibling trees rather than a single descendant tree. I'm interested in thinking about that more - but it's such a major shift in the meaning of 'scope' that it would require significant re-imagining, and some explicit switch between the two modes.  

Proximity, on the other hand, would be much simpler to extend. The current spec includes a `>>` [descendant combinator](https://www.w3.org/TR/css-cascade-6/#scope-combinator) that applies proximity weighting based on a descendant relationship, without any 'scope' behavior. I could imagine a similar `~~` proximity-weighted sibling combinator that applies proximity to sibling relationships. Since combinators already describe the desired direction of the relationship (descendant or sibling), there would not be confusion about their meaning. So Then your example would look like:

```css
h1 ~~ p {
	color: red;
}

h2 ~~ p {
	color: green;
}

h3 ~~ p {
	color: blue;
}
```

That would achieve what you're going for in the example - but it does have a slightly different meaning than applying a boundary. The main difference is in what happens with styles that _don't conflict_. If we add a margin to the first selector block, it would apply to all paragraphs following the h1, and not only the paragraphs between h1 and another header.

**tldr;** there's a pretty clear path for applying _cascade proximity_ to sibling selectors if we want it, but it would take some significant re-imagining of the `@scope` syntax to apply 'sibling boundaries'. I'm intrigued by the idea, but I'm not sure how well it actually maps onto existing definitions of 'scope' as a tree fragment with a single scope-root.

**Comment by romainmenke** [comment]
```css
h1 ~~ p {
	color: red;
}

h2 ~~ p {
	color: green;
}

h3 ~~ p {
	color: blue;
}
```

@mirisuzanne I really like this and I agree that only a proximity-weighted sibling combinator is needed to resolve my issue.

Thank you for your insights.

> In order to define sibling-based 'scopes', we would need a more major shift to the syntax.

Definitely want to avoid that :)

**Comment by mirisuzanne** [comment]
Agenda+ TPAC to see if we're happy with this path forward for proximity-weighted `++` and `~~` CSS combinators, in addition to the specified `>>` proximity-weighted descendant combinator.

**Comment by mirisuzanne** [comment]
Deferred this issue for Cascade 7, since the resolution is specific to combinators (deferred), and there is another open issue for block-level sibling scopes.
